Add a guide for the Elasticsearch Java API client

loicmathieu opened the pull request.

loicmathieu committed:
Add a guide for the Elasticsearch Java API client

loicmathieu commented:
Quickstart PR is here: https://github.com/quarkusio/quarkus-quickstarts/pull/1247

yrodiere commented on docs/src/main/asciidoc/elasticsearch.adoc:
```diff
@@ -103,7 +102,7 @@ public class Fruit {
 
 Now create a `org.acme.elasticsearch.FruitService` that will be the business layer of our application and store/load the fruits from the Elasticsearch instance.
-Here we use the low level client, if you want to use the high level client instead follow the instructions in the <<using-the-high-level-rest-client,Using the High Level REST Client>> paragraph instead.
+Here we use the low level client, if you want to use the Java APO client instead follow the instructions in the <<using-the-java-api-client,Using the Java API Client>> paragraph instead.
```
```suggestion
Here we use the low level client, if you want to use the Java API client instead follow the instructions in the <<using-the-java-api-client,Using the Java API Client>> paragraph instead.
```

yrodiere commented on docs/src/main/asciidoc/elasticsearch.adoc:
```diff
@@ -6,14 +6,13 @@ https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
 
-Quarkus provides two ways of accessing Elasticsearch: via the lower level `RestClient` or via the `RestHighLevelClient` we will call them
-the low level and the high level clients.
+Quarkus provides three ways of accessing Elasticsearch: via the lower level `RestClient`, via the high level `RestHighLevelClient`, or via the Java API client.
```
Unrelated to your PR, but shouldn't we just drop the high-level client?

It was deprecated in Quarkus 2, though of course the value of that deprecation is debatable since there wasn't any alternative...

@gsmet any opinion?

gsmet replied:
(forgot to post my review)

I agree with this, I will draft a proposal in a separate commit.

yrodiere commented on docs/src/main/asciidoc/elasticsearch.adoc:
```diff
@@ -66,21 +65,21 @@ For the Elasticsearch low level client, add:
 ----
 
-For the Elasticsearch high level client, add:
+For the Elasticsearch Java API client, add:
```
Unrelated to your PR, but this got me thinking that we should probably split the clients some time in the future. The guide doesn't quite flow naturally with all these "Or if you want this instead, do that".

yrodiere reviewed (commented):
I found a typo, but otherwise LGTM.

yrodiere commented:
Also, thanks for working on this :)

loicmathieu commented:
Agree with @gsmet we should clearly say that we will remove the high level client soon so I think the best is to remove all the section about it (or only keep the a warning explaining that we will drop support).
There is no point someone will start using it so there should not be instruction about it in the guide.

gsmet committed:
Clearly mark the high level REST client as deprecated

And removed the doc about it as we don't want people to use it anymore.
People interested in it can still have a look at the older documentation
if needed.

Also did some minor editorialization and formatting improvements.

gsmet commented:
@loicmathieu @yrodiere I pushed an additional commit, let me know what you think!

loicmathieu commented on docs/src/main/asciidoc/elasticsearch.adoc:
```diff
@@ -423,138 +437,25 @@ public class FruitService {
-2. We create an Elasticsearch index request using a builder.
-3. We directly pass the object to the request as the Java API client has a serialization layer.
-4. We send the request to Elasticsearch.
-
```
These lines are useful as they explain the code.
If you don't want the anymore, please remove the comments that reference these notes

yrodiere commented on docs/src/main/asciidoc/elasticsearch.adoc:
```diff
@@ -6,14 +6,24 @@ https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
+
+[WARNING]
+====
+The Quarkus extension for the high level REST Client is deprecated and will be removed in a future version,
```
Maybe clarify that the high level REST client is not the same as the "Java client" mentioned above? It could be confusing.

```suggestion
A third Quarkus extension for the "high level REST Client" exists, but is deprecated and will be removed in a future version,
```

yrodiere reviewed (commented):
Just a few suggestions, but feel free to merge.

gsmet commented on docs/src/main/asciidoc/elasticsearch.adoc:
```diff
@@ -6,14 +6,24 @@ https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
 include::_attributes.adoc[]
 :categories: data
-:summary: This guide covers how to use an Elasticsearch cluster using the low level or high level REST clients.
+:summary: This guide covers how to query an Elasticsearch cluster using the low level REST client or the Elasticsearch Java client.
```
Thanks for the suggestion, I wasn't happy with it!

gsmet commented:
@yrodiere applied your suggestions. Now begging for an approval :).

yrodiere commented:
> @yrodiere applied your suggestions. Now begging for an approval :).

Beg and you shall receive :P
